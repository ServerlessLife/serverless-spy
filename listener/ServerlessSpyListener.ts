import { PrettifyForDisplay } from './PrettifyForDisplay';
import {
  DynamoDBSpyHandler,
  EventBridgeSpyHandler,
  EventBridgeRuleSpyHandler,
  S3SpyHandler,
  SnsSubscriptionSpyHandler,
  SnsTopicSpyHandler,
  SqsSpyHandler,
  FunctionRequestSpyHandler,
  FunctionConsoleSpyHandler,
  FunctionResponseSpyHandler,
  FunctionErrorSpyHandler,
} from './SpyHandlers.ts';
import { WaitForParams } from './WaitForParams';
import { DynamoDBSpyEvent } from '../common/spyEvents/DynamoDBSpyEvent';
import { EventBridgeRuleSpyEvent } from '../common/spyEvents/EventBridgeRuleSpyEvent';
import { EventBridgeSpyEvent } from '../common/spyEvents/EventBridgeSpyEvent';
import { FunctionConsoleSpyEvent } from '../common/spyEvents/FunctionConsoleSpyEvent';
import { FunctionErrorSpyEvent } from '../common/spyEvents/FunctionErrorSpyEvent';
import { FunctionRequestSpyEvent } from '../common/spyEvents/FunctionRequestSpyEvent';
import { FunctionResponseSpyEvent } from '../common/spyEvents/FunctionResponseSpyEvent';
import { S3SpyEvent } from '../common/spyEvents/S3SpyEvent';
import { SnsSubscriptionSpyEvent } from '../common/spyEvents/SnsSubscriptionSpyEvent';
import { SnsTopicSpyEvent } from '../common/spyEvents/SnsTopicSpyEvent';
import { SqsSpyEvent } from '../common/spyEvents/SqsSpyEvent';

export type ServerlessSpyListener<TSpyEvents> = {
  [P in keyof FilterConditionally<TSpyEvents, `DynamoDB#${any}`> &
    string as `waitFor${P}`]: <T = any>(
    param?: PrettifyForDisplay<
      WaitForParams<PrettifyForDisplay<DynamoDBSpyEvent>>
    >
  ) => Promise<DynamoDBSpyHandler<T>>;
} & {
  [P in keyof FilterConditionally<TSpyEvents, `EventBridge#${any}`> &
    string as `waitFor${P}`]: <T = any>(
    param?: PrettifyForDisplay<
      WaitForParams<PrettifyForDisplay<EventBridgeSpyEvent<T>>>
    >
  ) => Promise<EventBridgeSpyHandler<T>>;
} & {
  [P in keyof FilterConditionally<TSpyEvents, `EventBridgeRule#${any}`> &
    string as `waitFor${P}`]: <T = any>(
    param?: PrettifyForDisplay<
      WaitForParams<PrettifyForDisplay<EventBridgeRuleSpyEvent<T>>>
    >
  ) => Promise<EventBridgeRuleSpyHandler<T>>;
} & {
  [P in keyof FilterConditionally<TSpyEvents, `S3#${any}`> &
    string as `waitFor${P}`]: (
    param?: PrettifyForDisplay<WaitForParams<PrettifyForDisplay<S3SpyEvent>>>
  ) => Promise<S3SpyHandler>;
} & {
  [P in keyof FilterConditionally<TSpyEvents, `SnsSubscription#${any}`> &
    string as `waitFor${P}`]: <T = any>(
    param?: PrettifyForDisplay<
      WaitForParams<PrettifyForDisplay<SnsSubscriptionSpyEvent<T>>>
    >
  ) => Promise<SnsSubscriptionSpyHandler<T>>;
} & {
  [P in keyof FilterConditionally<TSpyEvents, `SnsTopic#${any}`> &
    string as `waitFor${P}`]: <T = any>(
    param?: PrettifyForDisplay<
      WaitForParams<PrettifyForDisplay<SnsTopicSpyEvent<T>>>
    >
  ) => Promise<SnsTopicSpyHandler<T>>;
} & {
  [P in keyof FilterConditionally<TSpyEvents, `Sqs#${any}`> &
    string as `waitFor${P}`]: <T = any>(
    param?: PrettifyForDisplay<
      WaitForParams<PrettifyForDisplay<SqsSpyEvent<T>>>
    >
  ) => Promise<SqsSpyHandler<T>>;
} & {
  [P in keyof FilterConditionally<TSpyEvents, `Function#${any}#Request`> &
    string as `waitFor${P}`]: <T = any>(
    param?: PrettifyForDisplay<
      WaitForParams<PrettifyForDisplay<FunctionRequestSpyEvent<T>>>
    >
  ) => Promise<FunctionRequestSpyHandler<T>>;
} & {
  [P in keyof FilterConditionally<TSpyEvents, `Function#${any}#Console`> &
    string as `waitFor${P}`]: <T = any>(
    param?: PrettifyForDisplay<
      WaitForParams<PrettifyForDisplay<FunctionConsoleSpyEvent<T>>>
    >
  ) => Promise<FunctionConsoleSpyHandler<T>>;
} & {
  [P in keyof FilterConditionally<TSpyEvents, `Function#${any}#Error`> &
    string as `waitFor${P}`]: <T = any>(
    param?: PrettifyForDisplay<
      WaitForParams<PrettifyForDisplay<FunctionErrorSpyEvent<T>>>
    >
  ) => Promise<FunctionErrorSpyHandler<T>>;
} & {
  [P in keyof FilterConditionally<TSpyEvents, `Function#${any}#Response`> &
    string as `waitFor${P}`]: <T = any>(
    param?: PrettifyForDisplay<
      WaitForParams<PrettifyForDisplay<FunctionResponseSpyEvent<T>>>
    > //      filter?: (event: FunctionResponseSpyEvent<T>) => boolean
  ) => Promise<FunctionResponseSpyHandler<T>>;
} & {
  stop: () => void;
};

type FilterConditionally<Source, Condition> = Pick<
  Source,
  {
    [K in keyof Source]: Source[K] extends Condition ? K : never;
  }[keyof Source]
>;
