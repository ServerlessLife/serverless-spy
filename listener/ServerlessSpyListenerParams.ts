import { Credentials } from '@aws-sdk/types';

export type ServerlessSpyListenerParams = {
  credentials?: Credentials;
  debugMode?: boolean;
  scope: string;
};
