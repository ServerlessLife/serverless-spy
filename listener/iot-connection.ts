import { IoTClient, DescribeEndpointCommand } from '@aws-sdk/client-iot';
import { device } from 'aws-iot-device-sdk';

export const SSPY_TOPIC = 'sspy';

export type fragment = { id: string; index: number; count: number; data: any };

function createLog(debugMode: boolean) {
  return (message: string, ...optionalParams: any[]) => {
    if (debugMode) {
      console.debug('SSPY', message, ...optionalParams);
    }
  };
}

function createErrorLog() {
  return (message: string, ...optionalParams: any[]) => {
    console.error('SSPY', message, ...optionalParams);
  };
}

export async function getConnection(debugMode: boolean): Promise<device> {
  const log = createLog(debugMode);
  const logError = createErrorLog();
  log('Getting IoT endpoint');
  let response: any;
  try {
    const iotClient = new IoTClient({});
    response = await iotClient.send(
      new DescribeEndpointCommand({
        endpointType: 'iot:Data-ATS',
      })
    );
  } catch (e) {
    logError('failed to get endpoint', e);
    throw e;
  }
  log('Using IoT endpoint:', response.endpointAddress);

  if (!response.endpointAddress) {
    logError('No IoT endpoint could be found');
    throw new Error('IoT Endpoint address not found');
  }

  const connection = new device({
    protocol: 'wss',
    host: response.endpointAddress,
    region: process.env['AWS_REGION'],
    reconnectPeriod: 1,
  });

  connection.on('connect', () => {
    log('IoT connected');
  });

  connection.on('error', (err) => {
    logError('IoT error', err);
  });

  connection.on('close', () => {
    log('IoT closed');
  });

  connection.on('reconnect', () => {
    log('IoT reconnected');
  });

  return connection;
}
