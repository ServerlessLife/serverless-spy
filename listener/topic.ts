export const SSPY_TOPIC = 'sspy';

export function getTopic(scope: string) {
  return `${SSPY_TOPIC}/${scope}`;
}
