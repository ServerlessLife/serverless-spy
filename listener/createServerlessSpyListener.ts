import { ServerlessSpyListenerParams } from './ServerlessSpyListenerParams';
import { WsListener } from './WsListener';

export async function createServerlessSpyListener<TSpyEvents>(
  params: ServerlessSpyListenerParams
) {
  const wsListener = new WsListener<TSpyEvents>();
  let resolve, reject: ((value: void | PromiseLike<void>) => void) | undefined;
  const promise = new Promise<void>((res, rej) => {
    resolve = res;
    reject = rej;
  });
  await wsListener.start({
    ...params,
    connectionOpenResolve: params.connectionOpenResolve || resolve,
    connectionOpenReject: params.connectionOpenReject || reject,
  });

  const proxy = wsListener.createProxy();
  await promise;
  return proxy;
}
