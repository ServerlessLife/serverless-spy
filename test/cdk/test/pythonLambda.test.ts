import * as fs from 'fs';
import * as path from 'path';
import { LambdaClient, InvokeCommand } from '@aws-sdk/client-lambda';
import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { v4 as uuidv4 } from 'uuid';
import { createServerlessSpyListener } from '../../../listener/createServerlessSpyListener';
import { ServerlessSpyListener } from '../../../listener/ServerlessSpyListener';
import { ServerlessSpyEvents } from '../serverlessSpyEvents/ServerlessSpyEventsPythonLambda';
import { PythonLambdaStack } from '../src/pythonLambdaStack';
import { TestData } from './TestData';

jest.setTimeout(30000);

describe('PythonLambda', () => {
  const exportLocation = path.join(__dirname, '../cdkOutput.json');
  let serverlessSpyListener: ServerlessSpyListener<ServerlessSpyEvents>;
  let lambdaClient: LambdaClient;

  if (!fs.existsSync(exportLocation)) {
    throw new Error(`File ${exportLocation} does not exists.`);
  }
  const output = JSON.parse(fs.readFileSync(exportLocation).toString())[
    'ServerlessSpyPythonLambda'
  ];

  beforeEach(async () => {
    lambdaClient = new LambdaClient({});
    serverlessSpyListener =
      await createServerlessSpyListener<ServerlessSpyEvents>({
        scope: 'ServerlessSpyPythonLambda',
      });
  });

  afterEach(async () => {
    serverlessSpyListener.stop();
    lambdaClient.destroy();
  });

  test('Basic test', async () => {
    const myData1 = <TestData>{
      id: uuidv4(),
      message: 'Hello 1',
    };

    await lambdaClient.send(
      new InvokeCommand({
        FunctionName: output.FunctionNameMyLambda,
        InvocationType: 'RequestResponse',
        LogType: 'Tail',
        Payload: JSON.stringify(myData1) as any,
      })
    );

    await testWithoutConditionsAndWithoutChaining(
      serverlessSpyListener,
      myData1
    );

    await testWithConditionsAndWithoutChaining(serverlessSpyListener, myData1);
    await testWithConditionsAndWithChaining(serverlessSpyListener, myData1);
    await testFullyChained(serverlessSpyListener, myData1);

    //************** DATA 2 **************************

    const myData2 = <TestData>{
      id: uuidv4(),
      message: 'Hello 2',
    };

    await lambdaClient.send(
      new InvokeCommand({
        FunctionName: output.FunctionNameMyLambda,
        InvocationType: 'RequestResponse',
        LogType: 'Tail',
        Payload: JSON.stringify(myData2) as any,
      })
    );

    await testWithConditionsAndWithChaining(serverlessSpyListener, myData2);

    //this will not work, because we would get the first event
    // const requestData2 = await (
    //   await serverlessSpyListener.waitForFunctionMyLambdaRequest<TestData>()
    // ).getData();

    const followedResponse2 = await (
      await serverlessSpyListener.waitForFunctionMyLambdaRequest<TestData>({
        condition: (d) => d.request.id === myData2.id,
      })
    ).followedByResponse();

    followedResponse2.toMatchObject({
      request: myData2,
      response: {
        ...myData2,
        message: `${myData2.message} ServerlessSpy`,
      },
    });

    const responseDataSameRequest2 = followedResponse2.getData();

    expect(responseDataSameRequest2.request).toMatchObject(myData2);
    expect(responseDataSameRequest2.response).toMatchObject({
      ...myData2,
      message: `${myData2.message} ServerlessSpy`,
    });

    (
      await serverlessSpyListener.waitForFunctionMyLambdaResponse<TestData>({
        condition: (d) => d.response.id === myData2.id,
      })
    ).toMatchObject({
      request: myData2,
      response: {
        ...myData2,
        message: `${myData2.message} ServerlessSpy`,
      },
    });
  });

  test('Test error', async () => {
    const lambdaClient = new LambdaClient({});

    const myData1 = <TestData>{
      id: uuidv4(),
      message: 'Hello 1',
    };

    await lambdaClient.send(
      new InvokeCommand({
        FunctionName: output.FunctionNameMyLambdaThatFails,
        InvocationType: 'RequestResponse',
        LogType: 'Tail',
        Payload: JSON.stringify(myData1) as any,
      })
    );

    const errorResponse = (
      await serverlessSpyListener.waitForFunctionMyLambdaThatFailsError()
    ).getData();

    expect(errorResponse.error).toMatchObject({
      message: 'My test error',
      name: 'Error',
      stack: expect.any(String),
    });

    (
      await serverlessSpyListener.waitForFunctionMyLambdaThatFailsError()
    ).toMatchObject({
      request: myData1,
      error: {
        message: 'My test error',
        name: 'Error',
        stack: expect.any(String),
      },
    });
  });

  test('Snapshot', () => {
    const app = new App();
    const stack = new PythonLambdaStack(app, 'Test', {
      generateSpyEventsFile: false,
    });
    const template = Template.fromStack(stack);
    expect(template.toJSON()).toMatchSnapshot();
  });
});

async function testWithConditionsAndWithChaining(
  serverlessSpyListener: ServerlessSpyListener<ServerlessSpyEvents>,
  myData: TestData
) {
  // request
  const awaitedRequest =
    await serverlessSpyListener.waitForFunctionMyLambdaRequest<TestData>({
      condition: (d) => d.request.id === myData.id,
    });

  awaitedRequest.toMatchObject({
    request: myData,
  });

  const requestData = awaitedRequest.getData();

  expect(requestData.request).toMatchObject(myData);

  //console log
  const awaitedFollowedConsole = await awaitedRequest.followedByConsole();

  awaitedFollowedConsole.toMatchObject({
    request: myData,
    console: {
      message: 'My console log message',
      optionalParams: [myData],
    },
  });

  const consoleData = awaitedFollowedConsole.getData();

  expect(consoleData.console).toMatchObject({
    message: 'My console log message',
    optionalParams: [myData],
  });

  //response
  const awaitedFollowedResponse = await awaitedRequest.followedByResponse();

  awaitedFollowedResponse.toMatchObject({
    request: myData,
    response: {
      ...myData,
      message: `${myData.message} ServerlessSpy`,
    },
  });

  const responseData = awaitedFollowedResponse.getData();

  expect(responseData.request).toMatchObject(myData);
  expect(responseData.response).toMatchObject({
    ...myData,
    message: `${myData.message} ServerlessSpy`,
  });
}

async function testFullyChained(
  serverlessSpyListener: ServerlessSpyListener<ServerlessSpyEvents>,
  myData: TestData
) {
  (
    await (
      await (
        await serverlessSpyListener.waitForFunctionMyLambdaRequest<TestData>({
          condition: (d) => d.request.id === myData.id,
        })
      )
        .toMatchObject({
          request: myData,
        })
        .followedByConsole()
    )
      .toMatchObject({
        request: myData,
        console: {
          message: 'My console log message',
          optionalParams: [myData],
        },
      })
      .followedByResponse()
  ).toMatchObject({
    request: myData,
    response: {
      ...myData,
      message: `${myData.message} ServerlessSpy`,
    },
  });

  //no checking the data
  await (
    await (
      await serverlessSpyListener.waitForFunctionMyLambdaRequest<TestData>({
        condition: (d) => d.request.id === myData.id,
      })
    ).followedByConsole()
  ).followedByResponse();
}

async function testWithoutConditionsAndWithoutChaining(
  serverlessSpyListener: ServerlessSpyListener<ServerlessSpyEvents>,
  data: TestData
) {
  //request
  const awaitedRequest =
    await serverlessSpyListener.waitForFunctionMyLambdaRequest<TestData>();

  awaitedRequest.toMatchObject({
    request: data,
  });
  console.log('request found');

  const requestData = awaitedRequest.getData();

  expect(requestData.request).toMatchObject(data);
  console.log('matched data');

  //TODO: Pretty sure it's the console logs that aren't set up properly
  //console log
  const awaitedConsole =
    await serverlessSpyListener.waitForFunctionMyLambdaConsole<TestData>();

  awaitedConsole.toMatchObject({
    request: data,
    console: {
      message: 'My console log message',
      optionalParams: [data],
    },
  });

  const consoleData = awaitedConsole.getData();

  expect(consoleData.console).toMatchObject({
    message: 'My console log message',
    optionalParams: [data],
  });

  //response
  const awaitedResponse =
    await serverlessSpyListener.waitForFunctionMyLambdaResponse<TestData>();

  awaitedResponse.toMatchObject({
    request: data,
    response: {
      ...data,
      message: `${data.message} ServerlessSpy`,
    },
  });

  const responseData = awaitedResponse.getData();

  expect(responseData.request).toMatchObject(data);
  expect(responseData.response).toMatchObject({
    ...data,
    message: `${data.message} ServerlessSpy`,
  });
}

async function testWithConditionsAndWithoutChaining(
  serverlessSpyListener: ServerlessSpyListener<ServerlessSpyEvents>,
  data: TestData
) {
  //request
  const awaitedRequest =
    await serverlessSpyListener.waitForFunctionMyLambdaRequest<TestData>({
      condition: (d) => d.request.id === data.id,
    });

  awaitedRequest.toMatchObject({
    request: data,
  });

  const requestData = awaitedRequest.getData();

  expect(requestData.request).toMatchObject(data);

  //console log
  const awaitedConsole =
    await serverlessSpyListener.waitForFunctionMyLambdaConsole<TestData>({
      condition: (d) => d.request.id === data.id,
    });

  awaitedConsole.toMatchObject({
    request: data,
    console: {
      message: 'My console log message',
      optionalParams: [data],
    },
  });

  const consoleData = awaitedConsole.getData();

  expect(consoleData.console).toMatchObject({
    message: 'My console log message',
    optionalParams: [data],
  });

  //response
  const awaitedResponse =
    await serverlessSpyListener.waitForFunctionMyLambdaResponse<TestData>({
      condition: (d) => d.response.id === data.id,
    });

  awaitedResponse.toMatchObject({
    request: data,
    response: {
      ...data,
      message: `${data.message} ServerlessSpy`,
    },
  });

  const responseData = awaitedResponse.getData();

  expect(responseData.request).toMatchObject(data);
  expect(responseData.response).toMatchObject({
    ...data,
    message: `${data.message} ServerlessSpy`,
  });
}
