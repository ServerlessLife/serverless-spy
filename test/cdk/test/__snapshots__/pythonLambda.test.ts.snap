// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PythonLambda Snapshot 1`] = `
{
  "Outputs": {
    "FunctionNameMyLambda": {
      "Value": {
        "Ref": "MyLambdaCCE802FB",
      },
    },
    "FunctionNameMyLambdaThatFails": {
      "Value": {
        "Ref": "MyLambdaThatFails5E489081",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "MyLambdaCCE802FB": {
      "DependsOn": [
        "MyLambdaServiceRoleDefaultPolicy5BBC6F68",
        "MyLambdaServiceRole4539ECB6",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "a34102ca9b461fd6335e146271be5476c61e07b3b0ca21959ff0705dd70432d8.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_LAMBDA_EXEC_WRAPPER": "/opt/spy-wrapper",
            "SSPY_DEBUG": "true",
            "SSPY_FUNCTION_NAME": "MyLambda",
            "SSPY_INFRA_MAPPING": "{}",
            "SSPY_ROOT_STACK": "Test",
          },
        },
        "Handler": "lambda.handler",
        "Layers": [
          {
            "Ref": "ServerlessSpyPythonExtension718A6034",
          },
          {
            "Ref": "ServerlessSpySpyWrapperLayerBE5627BF",
          },
        ],
        "MemorySize": 512,
        "Role": {
          "Fn::GetAtt": [
            "MyLambdaServiceRole4539ECB6",
            "Arn",
          ],
        },
        "Runtime": "python3.9",
        "Timeout": 5,
      },
      "Type": "AWS::Lambda::Function",
    },
    "MyLambdaServiceRole4539ECB6": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "MyLambdaServiceRoleDefaultPolicy5BBC6F68": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "iot:*",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "MyLambdaServiceRoleDefaultPolicy5BBC6F68",
        "Roles": [
          {
            "Ref": "MyLambdaServiceRole4539ECB6",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "MyLambdaThatFails5E489081": {
      "DependsOn": [
        "MyLambdaThatFailsServiceRoleDefaultPolicyE947DE2E",
        "MyLambdaThatFailsServiceRole961B78F0",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "a34102ca9b461fd6335e146271be5476c61e07b3b0ca21959ff0705dd70432d8.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_LAMBDA_EXEC_WRAPPER": "/opt/spy-wrapper",
            "SSPY_DEBUG": "true",
            "SSPY_FUNCTION_NAME": "MyLambdaThatFails",
            "SSPY_INFRA_MAPPING": "{}",
            "SSPY_ROOT_STACK": "Test",
          },
        },
        "Handler": "lambdaFail.handler",
        "Layers": [
          {
            "Ref": "ServerlessSpyPythonExtension718A6034",
          },
          {
            "Ref": "ServerlessSpySpyWrapperLayerBE5627BF",
          },
        ],
        "MemorySize": 512,
        "Role": {
          "Fn::GetAtt": [
            "MyLambdaThatFailsServiceRole961B78F0",
            "Arn",
          ],
        },
        "Runtime": "python3.9",
        "Timeout": 5,
      },
      "Type": "AWS::Lambda::Function",
    },
    "MyLambdaThatFailsServiceRole961B78F0": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "MyLambdaThatFailsServiceRoleDefaultPolicyE947DE2E": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "iot:*",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "MyLambdaThatFailsServiceRoleDefaultPolicyE947DE2E",
        "Roles": [
          {
            "Ref": "MyLambdaThatFailsServiceRole961B78F0",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ServerlessSpyPythonExtension718A6034": {
      "Properties": {
        "CompatibleArchitectures": [
          "x86_64",
        ],
        "CompatibleRuntimes": [
          "python3.9",
        ],
        "Content": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "f0805f3798a1ec284ae50530bcfe1f7dc297a30367d2ef35819e3b21b57db6cd.zip",
        },
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "ServerlessSpySpyWrapperLayerBE5627BF": {
      "Properties": {
        "Content": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "bf2df0858ecef77ac31b1994b46cd25f69a84fea5d58e9b5e5c6e458c33012f6.zip",
        },
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "ServerlessSpySubscription0036DF4AB": {
      "DependsOn": [
        "ServerlessSpySubscription0ServiceRoleDefaultPolicyD8FA77F4",
        "ServerlessSpySubscription0ServiceRoleF955D76C",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "2fc02cc64fedb8a5c7c3049c58656172132663d36b541073ada7da4078957efa.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "NODE_OPTIONS": "--enable-source-maps",
            "SSPY_INFRA_MAPPING": "{}",
            "SSPY_ROOT_STACK": "Test",
          },
        },
        "Handler": "index.handler",
        "MemorySize": 512,
        "Role": {
          "Fn::GetAtt": [
            "ServerlessSpySubscription0ServiceRoleF955D76C",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 5,
      },
      "Type": "AWS::Lambda::Function",
    },
    "ServerlessSpySubscription0ServiceRoleDefaultPolicyD8FA77F4": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "iot:*",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ServerlessSpySubscription0ServiceRoleDefaultPolicyD8FA77F4",
        "Roles": [
          {
            "Ref": "ServerlessSpySubscription0ServiceRoleF955D76C",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ServerlessSpySubscription0ServiceRoleF955D76C": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
